Manual: RGASP alignment evaluation scripts
==========================================

This text describes how to use the alignment evaluation scripts
written for RGASP, starting from BAM files to obtain tables and plots
of results.

Please send any questions to par.engstrom@scilifelab.se.

Prerequisites
-------------

First, make sure the dependencies are installed (see README.md). Then
set the following two environment variables, preferably in your
~/.bash_profile (or the equivalent file if you are using a shell other
than bash):

- RGASP_ALI_HOME: Base directory of this package
- RGASP_ALI_DATA: Data directory to use (can be your own data or the sample data included)

For example:

    RGASP_ALI_HOME=~/rgasp/alignment
    RGASP_ALI_DATA=~/rgasp/alignment/data

Organization of scripts
-----------------------

The processing of alignment files to extract statistics is carried out
by a number of perl scripts. The repository also includes several
R scripts that read the output from the perl scripts, and summarize
the results in plots and tables.

Most of the perl scripts process a single alignment file, are
configured through command-line options and do not make
assumptions about the naming of files or their organization in the
filesystem. Exceptions are validate_bam.pl and cufflinks_eval.pl (but
we shall strive to generalize those further).

In contrast, the R scripts generally summarize results collected from
different alignment files to present them for comparison. These
scripts therefore use configuration files that list aligners and data
sets, and make assumptions about where data files are located. All
configuration and data files should be in subdirectories under the
path specified in $RGASP_ALI_DATA.

Configuration files
-------------------

The R scripts use three configuration files:

- $RGASP_ALI_DATA/aligners/methods.txt
- $RGASP_ALI_DATA/aligners/styles.txt
- $RGASP_ALI_DATA/reads/datasets.txt

### methods.txt ###

This file lists the evaluated alignment protocols.

- name: Symbolic name for protocol. May contain spaces.
- team: Name of developer team contributing the protocol. Should
  consist of word characters only (A-z, 0-9 and _).
- number: Protocol number (a team may contribute several protocols and each should have a unique number)
- description: Description of protocol (optional)
- color: Color used for plotting symbols associated with this protocol
- style: Plotting style used for this protocol (see styles.txt)

Note that the combination of _team_ and _number_ is used as a unique
ID for each protocol, and _name_ should also be unique.

The repository currently contains a file methods_all.txt with all
protocols evaluated in RGASP, as well as a file methods.txt with a
subset of protocols, for which data files are provided. (Following
publication of the manuscript we shall add data files for all
protocols.)

### styles.txt ###

Each protocol has an associated plotting style. Those associations are
defined in methods.txt, as described above.

The plotting styles are defined in styles.txt, using three columns:

- style: Numeric ID of style (corresponds to the number given in methods.txt)
- lty: Line type for this style (corresponds to lty parameter in R plot functions)
- pch: Plot symbol for this style (corresponds to pch parameter in R plot functions)

Most scripts will only consider protocols with style ID in the range
1-4. In RGASP, higher style IDs were used for exploratory purposes and
assigned to protocols not included among those evaluated.

### datasets.txt ###

This files lists the RNA-seq data sets used.

- id: ID for data set. Should consist of word characters only.
- name: Name of dataset. May contain spaces.
- fragments: Total number of read pairs in data set.

Data file locations
-------------------

Data files generated by the perl scripts are located under the
directory _data/output_ in the repository. These data files are read by the
R scripts that generate plots and tables.

Currently a subset of the data files from RGASP are provided for
demonstration purposes. We will add all relevant data files following
publication of the manuscript.

Running the scripts
-------------------

Brief descriptions of all scripts are provided in the file script_index.txt.

Run any of the perl scripts without arguments to see a usage message
briefly explaining what the script does and what command-line
parameters can be used. Exceptions are scripts that do not take
arguments; these will print a usage message if any arguments are given.

The R scripts do not take any arguments, and can be executed as:

    R --no-save --no-restore <$RGASP_ALI_HOME/scripts/R/script.R

(where script.R is the name of the script)

It does not matter what your current working directory is for the
functioning of the R scripts. They will find the input files using the
environment variables. However, the scripts will create their output
files in the current working directory, so you may want to create an
empty directory and run the scripts from there.

Most of the R scripts can be run using the sample data files included
under *data/output* in this repository. For this to work, you must set
the environment variable *RGASP_ALI_DATA* to the path to the 
directory *data* in this repository.

The following R scripts should run using the sample data if the
environment variables are correctly set:

- make_alignment_vs_cufflinks_gaps_tables.R
- make_multi-intron_accuracy_tables.R
- make_pair_ali_yield_table.R
- make_sim_eval_tables.R
- plot_intron_known_vs_novel.R
- plot_sim_indel_accuracy.R
- plot_sim_intron_roc.R
- plot_sim_read_placement_bars.R
- plot_sim_splice_accuracy_scatter.R
- plot_sim_splice_recall_by_coverage.R
